From a0f67f49e14790e908887e435d91ab594b775e87 Mon Sep 17 00:00:00 2001
From: Stefano Babic <stefano.babic@swupdate.org>
Date: Tue, 3 Oct 2023 18:06:24 +0200
Subject: [PATCH] libubootenv: support newer yaml config

libubootenv supports now a configuartion file in YAML.  Adjust initialization
to support the new format that requires a different initialization.

This is also a preparation to full support libubootenv's namespaces in
SWUpdate and to have a persistent area for variables handled by SWUpdate
but not by the bootloader.

Signed-off-by: Stefano Babic <stefano.babic@swupdate.org>
---
 bootloader/uboot.c | 18 ++++++++++++------
 1 file changed, 12 insertions(+), 6 deletions(-)

diff --git a/bootloader/uboot.c b/bootloader/uboot.c
index a2f0794..7fd47ac 100644
--- a/bootloader/uboot.c
+++ b/bootloader/uboot.c
@@ -39,14 +39,20 @@ static struct {
 
 static int bootloader_initialize(struct uboot_ctx **ctx)
 {
-	if (libuboot.initialize(ctx, NULL) < 0) {
-		ERROR("Error: environment not initialized");
-		return -ENODEV;
-	}
-	if (libuboot.read_config(*ctx, CONFIG_UBOOT_FWENV) < 0) {
-		ERROR("Configuration file %s wrong or corrupted", CONFIG_UBOOT_FWENV);
+	int ret;
+	const char *namespace = NULL;
+
+	ret = libuboot_read_config_ext(ctx, CONFIG_UBOOT_FWENV);
+	if (ret) {
+		ERROR("Cannot initialize environment from %s", CONFIG_UBOOT_FWENV);
 		return -EINVAL;
 	}
+
+	namespace = libuboot_namespace_from_dt();
+
+	if (namespace)
+		*ctx = libuboot_get_namespace(*ctx, namespace);
+
 	if (libuboot.open(*ctx) < 0) {
 		WARN("Cannot read environment, using default");
 		if (libuboot.load_file(*ctx, CONFIG_UBOOT_DEFAULTENV) < 0) {
